### ISSUE TITLE
useLazyQuerywith Debounce

### ISSUE THREAD

**Issue by AjithRithik** [apollographql/react-apollo#3830]
**Intended outcome:**

When I call the useLazyQuery function inside the useEffect hook with the dependency of debounceSearchInput. What I expect is, when I change the input value which affect the debouncesearchInput with the idle time of 800ms and execute the useLazyQuery's function inside the useEffect hook..

**Actual outcome:**
Actual outcome I got,  when I change the input value, every time the useLazyQuery's function inside the useEffect hook get execute, I don't know why? I also logged console in useEffect hook, the console log execute as i expected, but the problem is useLazyQuery's function execution.

**How to reproduce the issue:**
Use this codesandbox link to reproduce this issue (https://codesandbox.io/s/gracious-williams-lvx0r)

**Version**
package.json:
{
  "name": "samplegql",
  "version": "0.1.0",
  "private": true,
  "dependencies": {
    "@testing-library/jest-dom": "^4.2.4",
    "@testing-library/react": "^9.3.2",
    "@testing-library/user-event": "^7.1.2",
    "apollo-boost": "^0.4.7",
    "apollo-cache-inmemory": "^1.6.5",
    "apollo-client": "^2.6.8",
    "apollo-link-http": "^1.5.16",
    "dotenv": "^8.2.0",
    "graphql": "^14.6.0",
    "graphql-tag": "^2.10.1",
    "react": "^16.12.0",
    "react-apollo": "^3.1.3",
    "react-dom": "^16.12.0",
    "react-scripts": "3.3.0"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject"
  },
  "eslintConfig": {
    "extends": "react-app"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  }
}

**Comment by alexpermyakov** [comment]
Try to configure Apollo as it shown here: https://www.apollographql.com/docs/react/get-started/
Also, try to use useLazyQuery from '@apollo/react-hooks', it should work.
Few more notes:

- GraphQL endpoint returns an error
- A hook should begin with 'use' prefix, so it should be useDebounce, not Debounce.

**Comment by krvajal** [comment]
I dont think this is a problem in apollo but in you application code. Your `debouncedSearchTerm` is not behaving as you think it is.  Try using something like this 
tps://usehooks.com/useDebounce/ (just an example) to get a debounced value.

**Comment by AjithRithik** [comment]
@alexpermyakov and @krvajal  I've made the code as your suggestion.
I got the solution as I expected. 
But I am facing little confusion here, that is the **useLazyquery**  variable name. The useLazyQuery got fired for every keypress when I directly place the searchInput value as a variable name but it's working fine when I place the debounces term value. Can anyone clarify me what happened in code?

My Code sandbox link : https://codesandbox.io/s/holy-sun-57wpj

**Comment by alexpermyakov** [comment]
@krvajal @AjithRithik I think this is an apollo issue, might be a duplicate for #3505. Your code is fine now.
I've replaced everything with the latest @apollo-client, still happens: https://codesandbox.io/s/modern-wildflower-j4rcm

**Comment by alexpermyakov** [comment]
Ok, I've figured it out. Apollo is totally fine, the reason why you send a request on every render is because you send a new name variable everytime. 
Instead of `searchInput.toString().trim()` should be `debouncedSearchTerm`. Then everything will work.

**Comment by pedro-lb** [comment]
Hey there!

As said before, it's likely an issue within your code.

You can use [use-debounced-effect-hook](https://www.npmjs.com/package/use-debounced-effect-hook) to add a debounce effect to React's useEffect easily. 😃
